npm/pkg/dataplane/types.go

```diff
@@ -76,3 +78,70 @@ func (npmPod *updateNPMPod) updateIPSetsToRemove(setNames []*ipsets.IPSetMetadat
 		npmPod.IPSetsToRemove = append(npmPod.IPSetsToRemove, set.GetPrefixName())
 	}
 }
+
+type updatePodCache struct {
+	sync.Mutex
+	cache map[string]*updateNPMPod
+	// queue maintains the FIFO queue of the pods in the cache.
+	// This makes sure we handle issue 1729 with the same queue as the control plane.
+	// It also lets us update Pod ACLs in the same queue as the control plane so that
+	// e.g. the first Pod created is the first Pod to have proper connectivity.
+	queue           []string
+	initialCapacity int
+}
+
+// newUpdatePodCache creates a new updatePodCache with the given initial capacity for the queue
+func newUpdatePodCache(initialCapacity int) *updatePodCache {
+	return &updatePodCache{
+		cache:           make(map[string]*updateNPMPod),
+		queue:           make([]string, 0, initialCapacity),
+		initialCapacity: initialCapacity,
+	}
+}
+
+// enqueue adds a pod to the queue if necessary and returns the pod object used
+func (c *updatePodCache) enqueue(m *PodMetadata) *updateNPMPod {
+	pod, ok := c.cache[m.PodKey]
+	if !ok {
+		klog.Infof("[DataPlane] pod key %s not found in updatePodCache. creating a new obj", m.PodKey)
+
+		pod = newUpdateNPMPod(m)
+		c.cache[m.PodKey] = pod
+		c.queue = append(c.queue, m.PodKey)
+	}
+
+	return pod
+}
+
+// dequeue returns the first pod in the queue and removes it from the queue.
+// Caller should ensure the queue is not empty.
+// Otherwise, the following will occur: "panic: runtime error: index out of range [0] with length 0"
+func (c *updatePodCache) dequeue() *updateNPMPod {
+	pod := c.cache[c.queue[0]]
```

Comment:
check if 0th element is present before calling for it

Reply:
discussed offline. caller must lock updatePodCache and check isEmpty() before calling dequeue()